### PR TITLE
Added NRMSE metric

### PR DESCRIPTION
**Pull request by chetan51**
Fixes https://github.com/numenta/nupic/issues/2351.

### CODE REVIEW

**Comment by chetan51** [comment]
@scottpurdy 
cc: @ywcui1990

**Comment by chetan51** [comment]
@scottpurdy 
cc: @ywcui1990

File: nupic/frameworks/opf/metrics.py [hunk 1]
```diff
@@ -507,6 +509,36 @@ def aggregate(self, accumulatedError, historyBuffer, steps):
 
 
 
+class MetricNRMSE(MetricRMSE):
+  """
+      computes normalized root-mean-square error
```
Comment (scottpurdy) [inline nupic/frameworks/opf/metrics.py]: fix doc string:
https://www.python.org/dev/peps/pep-0257/#one-line-docstrings

File: nupic/frameworks/opf/metrics.py [hunk 2]
```diff
@@ -507,6 +509,36 @@ def aggregate(self, accumulatedError, historyBuffer, steps):
...
+  """
+  def __init__(self, *args, **kwargs):
+    super(MetricNRMSE, self).__init__(*args, **kwargs)
+    self.maxValue = float("-inf")
```
Comment (scottpurdy) [inline nupic/frameworks/opf/metrics.py]: is this right? maxValue is smallest possible float and minVal is largest?

File: tests/unit/nupic/frameworks/opf/opf_metrics_test.py [hunk 3]
```diff
@@ -47,6 +47,19 @@ def testRMSE(self):
...
+  def testNRMSE(self):
+    nrmse = getModule(MetricSpec("nrmse", None, None,
+{"verbosity" : OPFMetricsTest.VERBOSITY}))
+    gt = [9, 4, 5, 6]
+    p = [0, 13, 8, 3]
+    for i in xrange(len(gt)):
+      nrmse.addInstance(gt[i], p[i])
+    target = 1.342
+
+    self.assertTrue(abs(nrmse.getMetric()["value"]-target)\
```
Comment (scottpurdy) [inline tests/unit/nupic/frameworks/opf/opf_metrics_test.py]: Why did you break the line? It looks like it is under 80 characters without the break and it is inside parens so no need for backslash anyway

File: nupic/frameworks/opf/metrics.py (hunk 2 above)
Comment (chetan51) [inline nupic/frameworks/opf/metrics.py]: It starts off like this, and then gets updated in `accumulate`.

File: tests/unit/nupic/frameworks/opf/opf_metrics_test.py (hunk 3 above)
Comment (scottpurdy) [inline tests/unit/nupic/frameworks/opf/opf_metrics_test.py]: Use `assertAlmostEqual` instead of manually checking the deviation from the target

**Comment by scottpurdy** [comment]
done reviewing

File: tests/unit/nupic/frameworks/opf/opf_metrics_test.py (hunk 3 above)
Comment (chetan51) [inline tests/unit/nupic/frameworks/opf/opf_metrics_test.py]: I copied that from another test in this file. I'll clean it up.

File: nupic/frameworks/opf/metrics.py (hunk 2 above)
Comment (scottpurdy) [inline nupic/frameworks/opf/metrics.py]: oh ok makes sense

**Comment by chetan51** [comment]
Updated.

File: tests/unit/nupic/frameworks/opf/opf_metrics_test.py [hunk 4]
```diff
@@ -47,6 +47,19 @@ def testRMSE(self):
...
+  def testNRMSE(self):
+    nrmse = getModule(MetricSpec("nrmse", None, None,
+{"verbosity" : OPFMetricsTest.VERBOSITY}))
+    gt = [9, 4, 5, 6]
+    p = [0, 13, 8, 3]
+    for i in xrange(len(gt)):
+      nrmse.addInstance(gt[i], p[i])
+    target = 3.5866
+
+    self.assertTrue(abs(nrmse.getMetric()["value"]-target) <
```
Comment (scottpurdy) [inline tests/unit/nupic/frameworks/opf/opf_metrics_test.py]: `assertAlmostEqual`?

File: tests/unit/nupic/frameworks/opf/opf_metrics_test.py [hunk 5]
```diff
@@ -47,6 +47,19 @@ def testRMSE(self):
 < OPFMetricsTest.DELTA)
 
 
+  def testNRMSE(self):
+    nrmse = getModule(MetricSpec("nrmse", None, None,
+{"verbosity" : OPFMetricsTest.VERBOSITY}))
```
Comment (scottpurdy) [inline tests/unit/nupic/frameworks/opf/opf_metrics_test.py]: indentation

**Comment by scottpurdy** [comment]
done reviewing

**Comment by chetan51** [comment]
Cleaned up.

File: tests/unit/nupic/frameworks/opf/opf_metrics_test.py [hunk 6]
```diff
@@ -47,6 +47,18 @@ def testRMSE(self):
 < OPFMetricsTest.DELTA)
 
 
+  def testNRMSE(self):
+    nrmse = getModule(MetricSpec("nrmse", None, None,
+                                 {"verbosity" : OPFMetricsTest.VERBOSITY}))
+    gt = [9, 4, 5, 6]
+    p = [0, 13, 8, 3]
+    for i in xrange(len(gt)):
```
Comment (scottpurdy) [inline tests/unit/nupic/frameworks/opf/opf_metrics_test.py]: Might be simpler:
```
for i, j in zip(gt, p):
  nrmse.addInstance(i, j)
```

**Comment by scottpurdy** [comment]
:+1:

**Comment by ywcui1990** [comment]
Looks good to me. A note that using NRMSE shouldn't give different optimization result than the RMSE metric, as the normalization factor solely depend on the data and shouldn't affect the final best model. It is good to have this anyway so that we can monitor NRMSE during optimization.

**Pull request merged**